DESIGN.md

```diff
@@ -0,0 +1,362 @@
+pointers to get familiar with the code. As such, it contains three parts.
+
+1. The [first part][languages] describes how Pirouette supports several input
+   languages as well as two languages that are particularly relevant to us: [the
+   example language] and Plutus IR – [Plutus intermediate
+   representation].
+
+2. The [second part][transformations] describes the transformations that
+   Pirouette performs on the input language to get it in a shape ready for
+   symbolic execution.
+
+3. The [third part][symbolic engine] digs into what constitutes the heart of
+   Pirouette: its symbolic engine, and in particular the [symbolic evaluator]
+   and the [symbolic prover].
+
+Languages
+---------
+
+[languages]: #languages
+
+### Pirouette's genericity
+
+Pirouette can process any System F-like language.
+More precisely, users can come and define their instance of
+`LanguageBuiltins`, which is Pirouette's way of defining the constants, builtin
+terms and builtin types of a language. The definition of the classes can be
+found in [`Pirouette.Term.Syntax.Base`]. For a simple instantiation of
+a language, you might want to check [the example language]. For a more involved
+one, you might want to check [the support for Plutus IR][plutus intermediate
+representation].
+
+### The example language
+
+[the example language]: #the-example-language
+
+The example language is the only instance of a language included in Pirouette.
+It is a simple functional language supporting higher-order types with explicit
+type annotations, basic types (booleans, integers, strings) with their
+associated operators, algebraic datatypes and pattern-matching. You will
+find all the heavy work in [`Language.Pirouette.Example`].
+
+The example language also has a concrete syntax heavily inspired by Haskell.
+Thanks to quasi-quoters, it is easy to define a set of unordered value
+definitions for the `Ex`ample language:
+
+```haskell
+myProgram :: PrtUnorderedDefs Ex
+myProgram = [prog|
+  addone : Integer -> Integer
+  addone x = x + 1
+
+  addtwo : Integer -> Integer
+  addtwo x = x + 2
+|]
+```
+
+Now for a more involved example. Say we want to compute a sum of integers in a
+generic way. We can do that by writing a function `sum` taking a list of
+integers and returning the total. In order to do that, we first need a notion of
+lists, which we choose to do in a generic way. Here we go:
+
+```haskell
+data List a
+  = Nil : List a
+  | Cons : a -> List a -> List a
+```
+
+Computing the sum of the elements of a list requires traversing it, which we can
+also decide to do in a generic way with a fold over the list.
+
+```haskell
+foldr : forall a r . (a -> r -> r) -> r -> List a -> r
+foldr @a @r f e l =
+  match_List @a l @r
+    e
+    (\(x : a) (xs : List a) . f x (foldr @a @r f e xs))
+```
+
+Note how quantification in types has to be explicit, and the definition of the
+datatype `List` also creates the corresponding `match_List` function taking, in
```

Comment:
Maybe a reference about this style of doing case analysis would be helpful.

Reply:
Anything particular in mind or should we look for something?